<div align="center"><a href="https://www.nvidia.com/en-us/deep-learning-ai/education/"><img src="./images/DLI_Header.png"></a></div>

# Content Based Filters with Real Data

To test our Content-based Filters Knowledge, we'll be using a subsection of the [Amazon Review Dataset](https://nijianmo.github.io/amazon/index.html), specifically the Electronics dataset. There are a few more challenges working with real data such as data cleanliness, memory contraints, and  how to verify the model is ready for real world testing.

## Objectives
This notebook demonstrates:
* How to split data to train and test a recommender model
  * [1. Exploring the Data](#1.-Exploring-the-Data)
  * [2. Building the Datasets](#2.-Building-the-Datasets)
* How to build a Content-based filter for large datasets
  * [3. In-Place Content-Based Filters](#3.-In-Place-Content-Based-Filters)
    * [3.1 Finding the Total Number of Points Given per User](#3.1-Finding-the-Total-Number-of-Points-Given-per-User)
    * [3.2 Finding the Total Number of Points Given per User per Category](#3.2-Finding-the-Total-Number-of-Points-Given-per-User-per-Category)
    * [3.3 Finding the User's Percentage Preference for Each Category](#3.3-Finding-the-User's-Percentage-Preference-for-Each-Category)
  * [4. Making a Prediction](#4.-Making-a-Prediction)
  * [5. Validation Metrics](#5.-Validation-Metrics)
  * [6. Wrap Up](#6.-Wrap-Up)

## 1. Exploring the Data

Our data is located in a csv file under `/data`. We've already merged the metadata with the reviews data so everything can be displayed in one DataFrame.

In [1]:
import cudf

ratings = cudf.read_csv("data/raw_data.csv")

ratings.columns

Index(['reviewerID', 'asin', 'overall', 'unixReviewTime', 'brand',
       'category_0_0', 'category_0_1', 'category_0_2', 'category_0_3',
       'category_1_0', 'category_1_1', 'category_1_2', 'category_1_3',
       'salesRank_Electronics', 'salesRank_Camera', 'salesRank_Computers',
       'salesRank_CellPhones', 'salesRank_CellPhones_NA',
       'salesRank_Electronics_NA', 'salesRank_Camera_NA',
       'salesRank_Computers_NA', 'price_filled'],
      dtype='object')

Hmm, it looks like our categories are being stored as strings alongside the user-item ratings. Previously, we had a column for each category with a `1` if the item was that category, and `0` otherwise. Can we make that here? 

Let's start by getting a list of all the categories. We can [concatenate](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.concat.html) to stack all the category columns together, and then use [unique](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.Series.unique.html) to remove any duplicates.

In [2]:
category_columns = [
    'category_0_0', 'category_0_1', 'category_0_2', 'category_0_3',
    'category_1_0', 'category_1_1', 'category_1_2', 'category_1_3'
]

categories = cudf.Series(dtype="str")
for category in category_columns:
    categories = cudf.concat([categories, ratings[category]])
categories = categories.unique()

categories

0                       3D Glasses
1                      AC Adapters
2        AV Receivers & Amplifiers
3           Access-Control Keypads
4                      Accessories
                  ...             
581                       Xbox 360
582                       Xbox One
583                      Zip Discs
584                  eBook Readers
585    eBook Readers & Accessories
Length: 586, dtype: object

Wow! 586 categories! We could have a column for each category like before, but is that wise? How many rows does our data have?

In [3]:
count_ratings = len(ratings)

count_ratings

1689188

If we do create a column for each value, we'll have a billion cells worth of data. Okay, new strategy. We're going to  build our filter using only a few more extra columns with our dataset.

## 2. Building the Datasets

Before we get much further, we should split our dataset into training and validation datasets. One thing to consider when doing this for recommendation systems is that our goal is to encourage future interaction by the users. Instead of doing a random split, we'll use the last review left by the user for validation.

The `unixReviewTime` column enables us to do that. We will start by finding the largest (newest) review time and assigning it a marker, `valid`. However, the `unixReviewTime` only goes down to the day, so if a user does all of their reviews on the same day, then all of their reviews would end up in the validation set. We can add a tie-breaker, or `timeBreaker`, by adding a scaled version of the row index to the time. 

The `as_index` parameter for [groupby](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.DataFrame.groupby.html) will create a fresh numerical index as opposed to using the `reviewerID` as the index.

In [4]:
scaled_index = ratings.index / count_ratings
ratings["timeBreaker"] = scaled_index + ratings["unixReviewTime"]

In [5]:
valid_ratings = ratings[["reviewerID", "timeBreaker"]].groupby(['reviewerID'], as_index=False).max()
valid_ratings["valid"] = True

valid_ratings.head()

reviewerID   timeBreaker  valid
0  A000715434M800HLCENK9  1.400458e+09   True
1   A00101847G3FJTWYGNQA  1.385770e+09   True
2  A00166281YWM98A3SVD55  1.370563e+09   True
3  A0046696382DWIPVIWO0K  1.402877e+09   True
4  A00472881KT6WR48K907X  1.402358e+09   True

Next, we [merge](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.DataFrame.merge.html) the validation marker back onto our full ratings dataset to make it easier to pull the training data.

In [6]:
data_split = ratings.merge(valid_ratings, how="left", on=['reviewerID', "timeBreaker"])

data_split.head()

reviewerID        asin  overall  unixReviewTime brand category_0_0  \
0  A3V543DBS5C4RU  B0000510ZO      5.0      1245628800   C2G  Electronics   
1  A273KT0BQJCAXE  B0000510ZO      5.0      1374192000   C2G  Electronics   
2  A1RYZ3WCLIEZHK  B0000510ZO      1.0      1353283200   C2G  Electronics   
3  A1IWH0NZGT65RM  B0000510ZO      5.0      1395360000   C2G  Electronics   
4   ADE6RHHG0SWWM  B0000510ZO      4.0      1373587200   C2G  Electronics   

             category_0_1               category_0_2            category_0_3  \
0  Accessories & Supplies  Audio & Video Accessories  Cables & Interconnects   
1  Accessories & Supplies  Audio & Video Accessories  Cables & Interconnects   
2  Accessories & Supplies  Audio & Video Accessories  Cables & Interconnects   
3  Accessories & Supplies  Audio & Video Accessories  Cables & Interconnects   
4  Accessories & Supplies  Audio & Video Accessories  Cables & Interconnects   

  category_1_0  ... salesRank_Camera  salesRank_Computers  \
0           NA  ...          18665.5              14369.5   
1           NA  ...          18665.5              14369.5   
2           NA  ...          18665.5              14369.5   
3           NA  ...          18665.5              14369.5   
4           NA  ...          18665.5              14369.5   

   salesRank_CellPhones  salesRank_CellPhones_NA  salesRank_Electronics_NA  \
0              254050.0                    False                     False   
1              254050.0                    False                     False   
2              254050.0                    False                     False   
3              254050.0                    False                     False   
4              254050.0                    False                     False   

   salesRank_Camera_NA  salesRank_Computers_NA  price_filled   timeBreaker  \
0                False                   False          4.88  1.245629e+09   
1                False                   False          4.88  1.374192e+09   
2                False                   False          4.88  1.353283e+09   
3                False                   False          4.88  1.395360e+09   
4                False                   False          4.88  1.373587e+09   

   valid  
0   <NA>  
1   <NA>  
2   <NA>  
3   <NA>  
4   <NA>  

[5 rows x 24 columns]

To keep things clean, we'll replace the `null`s in the `valid` column with `False`.

In [7]:
data_split['valid'].fillna(False, inplace = True)

We'll keep a simplified table of the user-item validation ratings to make it easier to check our work later.

In [8]:
clean_columns = ["reviewerID", "asin", "overall"]
train_overall = data_split.loc[~data_split['valid']][clean_columns]

train_overall.head()

reviewerID        asin  overall
0  A3V543DBS5C4RU  B0000510ZO      5.0
1  A273KT0BQJCAXE  B0000510ZO      5.0
2  A1RYZ3WCLIEZHK  B0000510ZO      1.0
3  A1IWH0NZGT65RM  B0000510ZO      5.0
4   ADE6RHHG0SWWM  B0000510ZO      4.0

In [9]:
valid_overall = data_split.loc[data_split['valid']][clean_columns]

valid_overall.head()

reviewerID        asin  overall
10656   A94O1XOG5H69F  B000051WW9      3.0
10688  A2HNSWGBPZ2E3B  B00005RZYY      5.0
10720   AT2J7H5TRZM8Z  B0000659UN      4.0
10752  A2JMSJO1Z8C8ZQ  B00004SY4H      5.0
10753  A3MZSWSG6L8EN2  B00004SY4H      5.0

While we're at it, let's save some of this work for future notebooks.

In [10]:
save_columns = clean_columns + ['valid']

data_split[save_columns].to_csv('data/ratings.csv', index=False)

Here's the overall data efficiency strategy. We'll stack all the categories into one column, and then use `groupby` operations to get our user ratings for each category.

We'll again use the [concat](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.concat.html) function, but if we want all the categories to fall under the same column, we'll need to use the [rename](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.Series.rename.html) function to give each stack section the same category name.

A little more data cleanup... We'll also remove the `NA` category since it's an empty category, the `Electronics` category since every item has it, and any duplicate categories.

In [11]:
column_base = ["reviewerID", "asin", "overall", "valid"]

ratings_stack = cudf.DataFrame()
for category in category_columns:
    stack_section = data_split[column_base + [category]]
    stack_section = stack_section.rename(columns={category: "category"})
    ratings_stack = cudf.concat([ratings_stack, stack_section])

# Remove null categories and duplicates
ratings_stack = ratings_stack.loc[ratings_stack['category'] != "NA"]
ratings_stack = ratings_stack.loc[ratings_stack['category'] != "Electronics"]
ratings_stack = ratings_stack.drop_duplicates()

ratings_stack.head()

reviewerID        asin  overall  valid  \
325131  A000715434M800HLCENK9  B000UYYZ0M      1.0  False   
325131  A000715434M800HLCENK9  B000UYYZ0M      1.0  False   
325131  A000715434M800HLCENK9  B000UYYZ0M      1.0  False   
454864  A000715434M800HLCENK9  B001EHAI6Y      5.0  False   
454864  A000715434M800HLCENK9  B001EHAI6Y      5.0  False   

                              category  
325131          Accessories & Supplies  
325131  Office Electronics Accessories  
325131              Projection Screens  
454864          MP3 Player Accessories  
454864       MP3 Players & Accessories

Finally, we can use our `valid` marker and [loc](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.DataFrame.loc.html) to split our datasets.

In [12]:
train_data = ratings_stack.loc[~ratings_stack['valid']].drop("valid", axis=1)

train_data.head()

reviewerID        asin  overall  \
325131  A000715434M800HLCENK9  B000UYYZ0M      1.0   
325131  A000715434M800HLCENK9  B000UYYZ0M      1.0   
325131  A000715434M800HLCENK9  B000UYYZ0M      1.0   
454864  A000715434M800HLCENK9  B001EHAI6Y      5.0   
454864  A000715434M800HLCENK9  B001EHAI6Y      5.0   

                              category  
325131          Accessories & Supplies  
325131  Office Electronics Accessories  
325131              Projection Screens  
454864          MP3 Player Accessories  
454864       MP3 Players & Accessories

In [13]:
valid_data = ratings_stack.loc[ratings_stack['valid']].drop("valid", axis=1)

valid_data.head()

reviewerID        asin  overall             category
1685079  A000715434M800HLCENK9  B00HMZG3YS      5.0         Bags & Cases
1685079  A000715434M800HLCENK9  B00HMZG3YS      5.0       Camera & Photo
1685079  A000715434M800HLCENK9  B00HMZG3YS      5.0         Camera Cases
1538432   A00101847G3FJTWYGNQA  B00B19L8LO      4.0       Computer Cases
1538432   A00101847G3FJTWYGNQA  B00B19L8LO      4.0  Computer Components

## 3. In-Place Content-Based Filters


Now that we have our user (`reviewerID`), item (`asin`), rating (`overall`), and category all lined up, we have the necessary ingredients to make a content-based filter.

The following is a puzzle. Please replace the `FIXME`s for each code block in the section. Each `FIXME` is a single string or method. Use the <a href="1-03_content_based_intro.ipynb">previous lab</a> as a hint, and if more help is needed, take a look at the solutions folder.

### 3.1 Finding the Total Number of Points Given per User


In [15]:
user_total = (
    train_data[["reviewerID", "overall"]].groupby(["reviewerID"], as_index=False).sum()
)

user_total.head()

reviewerID  overall
0  A000715434M800HLCENK9     33.0
1   A00101847G3FJTWYGNQA     67.0
2  A00166281YWM98A3SVD55     58.0
3  A0046696382DWIPVIWO0K     43.0
4  A00472881KT6WR48K907X     84.0

### 3.2 Finding the Total Number of Points Given per User per Category

In [19]:
category_total = (
    train_data[["reviewerID", "overall", "category"]]
    .groupby(["reviewerID", "category"], as_index=False)
    .sum()
)

category_total.head()

reviewerID                   category  overall
0  A000715434M800HLCENK9     Accessories & Supplies      6.0
1  A000715434M800HLCENK9  Audio & Video Accessories      2.0
2  A000715434M800HLCENK9     Cables & Interconnects      2.0
3  A000715434M800HLCENK9     MP3 Player Accessories      5.0
4  A000715434M800HLCENK9  MP3 Players & Accessories      5.0

### 3.3 Finding the User's Percentage Preference for Each Category

In [22]:
category_total = category_total.merge(user_total, how="left", on='reviewerID')

category_total.head()

reviewerID                   category  overall_x  overall_y
0  A10DEX93MX7GVW  MP3 Players & Accessories       10.0       55.0
1  A10DEX93MX7GVW     Portable Audio & Video       10.0       55.0
2  A10DFH2Y5LWEU0                Accessories       13.0      276.0
3  A10DFH2Y5LWEU0     Accessories & Supplies       14.0      276.0
4  A10DFH2Y5LWEU0  Audio & Video Accessories        9.0      276.0

In [23]:
category_total["ratio"] = category_total["overall_x"] / category_total["overall_y"]

category_total.head()

reviewerID                   category  overall_x  overall_y     ratio
0  A10DEX93MX7GVW  MP3 Players & Accessories       10.0       55.0  0.181818
1  A10DEX93MX7GVW     Portable Audio & Video       10.0       55.0  0.181818
2  A10DFH2Y5LWEU0                Accessories       13.0      276.0  0.047101
3  A10DFH2Y5LWEU0     Accessories & Supplies       14.0      276.0  0.050725
4  A10DFH2Y5LWEU0  Audio & Video Accessories        9.0      276.0  0.032609

In [24]:
category_total = category_total.drop(["overall_x", "overall_y"], axis=1)

category_total.head()

reviewerID                   category     ratio
0  A10DEX93MX7GVW  MP3 Players & Accessories  0.181818
1  A10DEX93MX7GVW     Portable Audio & Video  0.181818
2  A10DFH2Y5LWEU0                Accessories  0.047101
3  A10DFH2Y5LWEU0     Accessories & Supplies  0.050725
4  A10DFH2Y5LWEU0  Audio & Video Accessories  0.032609

## 4. Making a Prediction

Congrats on making the users' profiles. Let's use them to make a prediction! We'll start by [merging](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.DataFrame.merge.html) our predictions with our validation data. We'll use an `inner` merge to avoid including predictions outside our validation dataset.

We do not need the validation ratings just yet, but we do need the categories for each item in our validation set. This allows us to pull the relevant categories from the user profile to find how well their category tastes align with the categories of the item we're predicting for.

In [25]:
prediction = category_total.merge(
    valid_data, how="inner", on=["reviewerID", "category"]
)

prediction.head()

reviewerID                 category     ratio        asin  overall
0  A10UHQH1YL5Q6B               Headphones  0.046041  B00FT34ROS      5.0
1  A10VB212O1V4J3              Accessories  0.030534  B00COYOAYW      4.0
2  A100UD67AHFODS  Computers & Accessories  0.130855  B00HHRP11C      5.0
3  A10M1CEJIUXKVA          Travel Chargers  0.036496  B00CEZBKTO      5.0
4  A1134QXO79NG7B              Accessories  0.024793  B00AANQLRI      4.0

In [26]:
prediction = prediction.drop(["category", "overall"], axis=1).groupby(['reviewerID', 'asin'], as_index=False).sum()

prediction.head()

reviewerID        asin     ratio
0   A00101847G3FJTWYGNQA  B00B19L8LO  0.432836
1  A00166281YWM98A3SVD55  B007B5S8BU  0.344828
2  A0046696382DWIPVIWO0K  B00HMXIKCS  0.279070
3  A00472881KT6WR48K907X  B00GYL9KK0  0.059524
4  A00473363TJ8YSZ3YAGG9  B003NR57BY  0.395833

## 5. Validation Metrics

The best way to test a recommender model is to do [A/B testing](https://en.wikipedia.org/wiki/A/B_testing), as the goal is to maximize user interactivity with our systems. However, we do not want to risk turning tested users away with a bad model, so we can sanity check our model against other metrics beforehand.

Here, we'll calculate [Root Mean Squared Error](https://en.wikipedia.org/wiki/Root-mean-square_deviation) with our validation set. Let's merge our predicted ratings with the validation data's true rating.

In [27]:
prediction = prediction.merge(valid_overall, how="inner", on=['reviewerID', 'asin'])

prediction.head()

reviewerID        asin     ratio  overall
0  A2FFEOMU480KU1  B0009YJXMS  0.222222      5.0
1   AZV1QKEUNHBTK  B000AV4MB2  0.112000      5.0
2   AL18XFR6SP8VI  B000AB286S  0.083333      5.0
3  A1QX9VE4U6FQMM  B000AB286S  0.029412      3.0
4   AV1XMNYXW5H1R  B000AB286S  0.181818      5.0

As in our example problem [in the previous notebook](1-03_content_based_intro.ipynb), we'll need to scale up our predictions.

Please replace the `FIXME`s so that we have predicted ratings with the correct scale.

In [28]:
min_rating = 1
max_rating = 5
prediction["predicted_rating"] = prediction["ratio"] * (max_rating - min_rating) + min_rating

prediction.head()

reviewerID        asin     ratio  overall  predicted_rating
0  A2FFEOMU480KU1  B0009YJXMS  0.222222      5.0          1.888889
1   AZV1QKEUNHBTK  B000AV4MB2  0.112000      5.0          1.448000
2   AL18XFR6SP8VI  B000AB286S  0.083333      5.0          1.333333
3  A1QX9VE4U6FQMM  B000AB286S  0.029412      3.0          1.117647
4   AV1XMNYXW5H1R  B000AB286S  0.181818      5.0          1.727273

Next, we'll calculate the squared error by subtracting the true value from the predicted value and squaring the result.

In [29]:
prediction["squared_error"] = (
    prediction["predicted_rating"] - prediction["overall"]
) ** 2

prediction.head()

reviewerID        asin     ratio  overall  predicted_rating  \
0  A2FFEOMU480KU1  B0009YJXMS  0.222222      5.0          1.888889   
1   AZV1QKEUNHBTK  B000AV4MB2  0.112000      5.0          1.448000   
2   AL18XFR6SP8VI  B000AB286S  0.083333      5.0          1.333333   
3  A1QX9VE4U6FQMM  B000AB286S  0.029412      3.0          1.117647   
4   AV1XMNYXW5H1R  B000AB286S  0.181818      5.0          1.727273   

   squared_error  
0       9.679012  
1      12.616704  
2      13.444444  
3       3.543253  
4      10.710744

Finally, we can find the Root Mean Squared Error (RMSE) by [averaging](https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.DataFrame.mean.html) the `squared_error`s and taking the square root.

In [30]:
prediction["squared_error"].mean() ** .5

2.6019044951648507

So, roughly on average, the predictions are off by about 2 points. Considering the scale of ratings goes from 1 - 5, that's a pretty big error.

However, there are some things to note.
* Predicted ratings are relative to other items. Take the example where a user would rate item A 5 and item B 1. It doesn't matter so much that our system predicts {A 4, B 3} or {A 3, B 2.5}. In both cases, item A is recommended to the user over B.
* In the next notebook, we'll go over collaborative filtering, which can produce better predictions with more data. There are some instances where content-based filtering can make predictions where collaborative filtering cannot. Try to keep track of the differences while going through the next notebook.

## 6. Wrap Up

Congratulations on making it this far! Please run the cell below to shut down the kernel before moving on to the <a href="1-05_als.ipynb">next notebook</a>.

In [31]:
import IPython

app = IPython.Application.instance()
app.kernel.do_shutdown(True)

{'status': 'ok', 'restart': True}

<div align="center"><a href="https://www.nvidia.com/en-us/deep-learning-ai/education/"><img src="./images/DLI_Header.png"></a></div>